Web Extensions: Lets extensions move bookmark nodes across folders using the Bookmarks API

akutira-umich opened the pull request:
#### 5797be28109be4ea7ace0a0e9ec41b788c36c680
<pre>
Web Extensions: Lets extensions move bookmark nodes across folders using the Bookmarks API
<a href="https://rdar.apple.com/152506123">rdar://152506123</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=297119">https://bugs.webkit.org/show_bug.cgi?id=297119</a>

Reviewed by Brian Weinstein.

Adds bookmarks.move() to the WebProcess and UIProcess. Includes tests with error checking

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::WebExtensionContext::bookmarksMove):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::move):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureMoveBookmarksDelegate):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIMove)):
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate _webExtensionController:moveBookmarkWithIdentifier:toParent:atIndex:forExtensionContext:completionHandler:]):

Canonical link: <a href="https://commits.webkit.org/298596@main">https://commits.webkit.org/298596@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/45141a4e867080b547eed5d144afb5558a35f9dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115980 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35634 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26170 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122029 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66518 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117869 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36340 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44221 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88108 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118928 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28993 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104069 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68513 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28114 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22178 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65699 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98382 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22312 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125182 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42867 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32177 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96856 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43233 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100257 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96640 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24593 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41898 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19775 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38796 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42756 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48353 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42223 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45557 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43930 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -464,7 +464,46 @@
+    if (![controllerDelegate respondsToSelector:@selector(_webExtensionController:moveBookmarkWithIdentifier:toParent:atIndex:forExtensionContext:completionHandler:)]) {
+        completionHandler(toWebExtensionError(apiName, nullString(), @"it is not implemented"));
+            return;
+        }
```
The indentation got a little bit wonky here.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -464,7 +464,46 @@
+            return;
+        }
+        if (!movedBookmark) {
+            completionHandler(toWebExtensionError(apiName, nullString(), @"moved bookmark has failed"));
```
Nit: `moving bookmark failed` would be more readable here.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -275,7 +275,44 @@
+
+    if (![destination isKindOfClass:[NSDictionary class]]) {
+        if (outExceptionString)
+            *outExceptionString = toErrorString(nullString(), @"destiantion", @"property must be an object.").createNSString().autorelease();
```
Typo in `destiantion`.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -275,7 +275,44 @@
+                return;
+            }
+
+            NSDictionary* movedNodeDictionary = toAPI(result.value());
```
Super-nit: The `*` should be on the other side of the space. `NSDictionary *movedNodeDIctionary`.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -413,6 +413,64 @@ void configureUpdateBookmarksDelegate(TestWebExtensionManager *manager)
+                newParentChildrenArray = uiProcessMockBookmarks.get();
+            else {
+                NSMutableDictionary *newParent = findBookmarkInMockTree(uiProcessMockBookmarks.get(), toParentId);
+                if (!newParent || ![newParent[@"type"] isEqualToString:@"folder"] || !newParent[@"children"]) {
```
What does the `!newParent[@"children"]` check accomplish? Does this mean you couldn't move a bookmark into an empty folder?

b-weinstein replied:
Should this method create an empty folder if it needs to?

akutira-umich replied:
This means that the newParent doesn't have a children in its dictionary. If it was empty then this would not be nil. Since bookmark is the only type that would have its children as nil, and I already check for not a folder, this 3rd parameter check is unnecessary now.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -413,6 +413,64 @@ void configureUpdateBookmarksDelegate(TestWebExtensionManager *manager)
+                newParentChildrenArray = newParent[@"children"];
+            }
+
+            NSInteger rawTargetIndex = atIndex ? [atIndex integerValue] : newParentChildrenArray.count;
```
Nit: `atIndex.integerValue`.

Should this be an unsigned int?

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -413,6 +413,64 @@ void configureUpdateBookmarksDelegate(TestWebExtensionManager *manager)
+            if (oldParentChildrenArray == newParentChildrenArray && oldIndex < (NSUInteger)rawTargetIndex)
+                rawTargetIndex--;
+
+            NSInteger finalTargetIndex = rawTargetIndex;
```
If this was made an unsigned integer, you wouldn't need the first check on line 454/455.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -413,6 +413,64 @@ void configureUpdateBookmarksDelegate(TestWebExtensionManager *manager)
+                finalTargetIndex = newParentChildrenArray.count;
+
+            bookmarkToMove[@"parentId"] = toParentId ?: @"0";
+            NSInteger targetIndex = atIndex ? [atIndex integerValue] : newParentChildrenArray.count;
```
Ditto here about using an unsigned integer.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -413,6 +413,60 @@ void configureUpdateBookmarksDelegate(TestWebExtensionManager *manager)
+                newParentChildrenArray = newParent[@"children"];
+            }
+
+            NSUInteger rawTargetIndex = atIndex ? atIndex.integerValue : newParentChildrenArray.count;
```
This should use `unsignedIntegerValue`. If this was passed a negative number we would probably end up throwing an exception.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -413,6 +413,60 @@ void configureUpdateBookmarksDelegate(TestWebExtensionManager *manager)
+                finalTargetIndex = newParentChildrenArray.count;
+
+            bookmarkToMove[@"parentId"] = toParentId ?: @"0";
+            NSUInteger targetIndex = atIndex ? [atIndex integerValue] : newParentChildrenArray.count;
```
`unsignedIntegerValue` and dot syntax.

akutira-umich committed:
Web Extensions: Lets extensions move bookmark nodes across folders using the Bookmarks API
rdar://152506123
https://bugs.webkit.org/show_bug.cgi?id=297119

Reviewed by Brian Weinstein.

Adds bookmarks.move() to the WebProcess and UIProcess. Includes tests with error checking

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::WebExtensionContext::bookmarksMove):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::move):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureMoveBookmarksDelegate):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIMove)):
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate _webExtensionController:moveBookmarkWithIdentifier:toParent:atIndex:forExtensionContext:completionHandler:]):

Canonical link: https://commits.webkit.org/298596@main
